### PR TITLE
[TRTLLM-14555][perf] Optimize parallel VAE halo exchange path

### PR DESCRIPTION
**Pull request by taianz-nv**
## Description

Jira: [TRTLLM-14555](https://jirasw.nvidia.com/browse/TRTLLM-14555) (Task under epic [TRTLLM-12775](https://jirasw.nvidia.com/browse/TRTLLM-12775))

This PR consolidates the correctness and performance work for the spatially
parallel VAE halo path into one reviewable change. The three stages form one
end-to-end contract:

1. **Address the correct peer.** `HaloExchangeConv2dStride2` selects the
   adjacent process group, maps its peer to the global rank, and passes the
   matching group to PyTorch distributed. This fixes offset VAE groups such as
   global ranks `[2, 3]`, which could previously contact the wrong peer or hang.
   Each wrapper now validates and resolves the adjacent groups it will use at
   construction time, before inference begins.
2. **Preserve the physical layout.** Channels-last inputs materialize halo
   exchange buffers in contiguous NHWC/NDHWC order and concatenate along the
   corresponding physical spatial axis. The result remains channels-last, so
   downstream convolutions avoid full-tensor layout conversions. Row-major
   inputs retain the existing path.
3. **Produce only the local output.** Safe native Wan stride-1 halo
   convolutions use zero padding on the split axis. For a local width of 20 and
   one element of halo on each side, this changes
   `W=22, pad=1 -> W=22 -> strip` into `W=22, pad=0 -> W=20`. The fast path is
   disabled when its split-axis geometry preconditions are not met. Diffusers
   modules retain their existing convolution-plus-strip implementation.

The implementation supports 2D and 3D convolutions and height and width splits.
It contains no FP4-specific code and introduces no public API or environment
variable.

## Performance

Four-GPU NSC measurements from the staged Wan VAE optimization campaign:

| Stage | Measured effect |
| --- | ---: |
| Physical halo concatenation | `-54.663 ms` VAE-only latency |
| Contiguous NHWC/NDHWC exchange buffers | further `-25.131 ms` VAE-only latency |
| Direct local-width Conv output | `55.364 -> 52.426 ms` across 360 isolated calls (`-2.938 ms`, `5.31%`) |

The first two entries are consecutive points in the four-GPU VAE-only ladder.
The direct-output number uses an isolated exact-shape Conv sweep and is listed
separately rather than summed with the ladder values.

## Test Coverage

- Offset-subgroup stride-2 parity verifies that VAE-local ranks map to the
  correct global peers.
- Distributed 2D/3D tests cover height/width splits, row-major inputs,
  `channels_last` / `channels_last_3d` inputs, and asymmetric even-kernel halo
  trimming.
- CPU tests cover halo exchange-buffer round trips, physical concatenation,
  eager adjacent-group validation, and the kernel-1 no-communication case.
- Native Wan tests cover direct local-size output for height and width splits
  and verify that unsafe split geometry does not select that optimization.
- Four-GPU NSC array job `1554035` tasks 0-2 passed the original distributed
  test groups.
- The consolidated diff passes repository pre-commit hooks and
  `git diff --check`.

## PR Checklist

- [x] Correctness and performance motivation are documented.
- [x] The change is precision-independent and contains no FP4 dependency.
- [x] Row-major and Diffusers paths are retained.
- [x] No public API, dependency, CODEOWNERS, or architecture-diagram changes.

## Dev Engineer Review

- Optimizes 2D and 3D halo exchange for channels-last tensors.
- Resolves global peer ranks and validates adjacent process groups.
- Preserves NHWC and NDHWC layouts during halo packing, exchange, trimming, and concatenation.
- Adds local-output optimization for supported stride-1 Wan convolutions.
- Preserves fallback behavior for unsupported geometries.
- Extends `WanCausalConv3d.forward` with the optional `spatial_padding` argument.
- Adds the Wan halo test to the B200 pre-merge test list.
- No public API or environment-variable changes are required.

## QA Engineer Review

- Added tests for channels-last 2D and 3D halo exchange.
- Added tests for even-kernel 3D exchange and offset-group stride-2 exchange.
- Added validation tests for adjacent process groups.
- Added tests for physical-layout concatenation and channel-last buffer round trips.
- Added tests for native Wan local-output optimization and unsupported-geometry fallback.
- `tests/unittest/_torch/visual_gen/test_wan_vae_halo.py` is listed in `tests/integration/test_lists/test-db/l0_b200.yml`.
- The added multi-GPU convolution tests are not explicitly listed in the provided test-list change.
- Verdict: needs follow-up to confirm CI coverage for all added multi-GPU tests.

## FP4 integration report

This precision-independent optimization composes with the FP4 Wan VAE integration in #17262. The consolidated performance/accuracy report and dependency map are maintained there.

### CODE REVIEW

File: tensorrt_llm/_torch/visual_gen/models/wan/parallel_vae.py [hunk 1]
```diff
@@ -80,13 +80,64 @@ def _native_decode_chunk_size(
...
+        # ``chunk_dim`` indexes NCTHW, while ``spatial_padding`` indexes HW.
+        conv_axis = self.chunk_dim - 2
+        spatial_axis = self.chunk_dim - 3
+        if (
+            self.module.stride[conv_axis] != 1
+            or self.module.dilation[conv_axis] != 1
+            or self.module.kernel_size[conv_axis] % 2 != 1
+            or self.module.padding[conv_axis] != self.halo_left
+            or self.module.padding[conv_axis] != self.halo_right
+        ):
```
Comment (zhenhuaw-me) [inline tensorrt_llm/_torch/visual_gen/models/wan/parallel_vae.py:113]: In my first impression, we only need to check the padding with hello in this context, since it should be a very general approach based on mathematic. But it should be fine, since the implementation is Wan VAE.

Comment (taianz-nv) [reply tensorrt_llm/_torch/visual_gen/models/wan/parallel_vae.py:113]: Agreed. For current Wan geometry, padding/halo matching is effectively the key condition; the extra checks make the stride-1 and dilation-1 assumptions explicit and keep the fast path fail-closed.

File: tests/unittest/_torch/visual_gen/test_wan_vae_halo.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+"""Unit tests for native Wan VAE halo convolution geometry."""
+
+from unittest import mock
+
+import pytest
+import torch
+
+from tensorrt_llm._torch.visual_gen.models.wan.parallel_vae import WanCausalConvHalo
+from tensorrt_llm._torch.visual_gen.models.wan.wan_vae import WanCausalConv3d
```
Comment (zhenhuaw-me) [inline tests/unittest/_torch/visual_gen/test_wan_vae_halo.py:12]: maybe move to the `multi_gpu` dir?

Comment (taianz-nv) [reply tests/unittest/_torch/visual_gen/test_wan_vae_halo.py:12]: This is intentionally a single-process test: it mocks both the process group and halo exchange to validate only the local padding/fast-path selection. Actual distributed communication remains covered in `multi_gpu/test_parallel_conv.py`, so I’ll keep this test in the single-GPU directory.
